### PR TITLE
Port `trace` cpu from TH to ATen

### LINKED ISSUE #24779: Migrate `trace` from the TH to Aten (CPU)
Porting TH operators is essential for code simplicity and performance reasons.

Porting guides and Q&A are available in umbrella issue: #24507

Feel free to add @VitalyFedyunin as a reviewer to get a prioritized review.

### PR DESCRIPTION
**Pull request by shihongzhi**
Fixes: #24779

### CODE REVIEW

**Comment by zou3519** [comment]
> @pytorchbot retest this please

I'm not sure this is working right now, you'll have to rebase the branch locally and push

**Comment by shihongzhi** [comment]
> > @pytorchbot retest this please
> 
> I'm not sure this is working right now, you'll have to rebase the branch locally and push

It's ready for review now.

File: aten/src/ATen/native/TensorShape.cpp [hunk 1]
```diff
@@ -1465,4 +1465,30 @@ Tensor& diag_out(Tensor &result, const Tensor& self, int64_t dimension) {
...
+Tensor trace(const Tensor& self) {
+  Tensor result = at::empty({}, self.options());
+  AT_DISPATCH_ALL_TYPES(self.scalar_type(), "trace", [&] {
+    scalar_t sum = 0;
+    auto t_data = self.data_ptr<scalar_t>();
+
+    int64_t i = 0;
+    int64_t t_stride_0, t_stride_1, t_diag_size;
+
+    TORCH_CHECK(self.dim() == 2, "expected a matrix");
```
Comment (zou3519) [inline aten/src/ATen/native/TensorShape.cpp]: It would be nice to make this error message more descriptive, something like "trace: expected a matrix, but got tensor with dim X"

**Review by zou3519 — COMMENTED** [review]
The code looks good. Could you provide a before/after performance comparison of trace on tensors with the following sizes:
- 1x1 (small)
- 1024x1024
(and any others you might want)

File: aten/src/ATen/native/TensorShape.cpp [hunk 2]
```diff
@@ -1465,4 +1465,30 @@ Tensor& diag_out(Tensor &result, const Tensor& self, int64_t dimension) {
   return result;
 }
 
+Tensor trace(const Tensor& self) {
+  Tensor result = at::empty({}, self.options());
+  AT_DISPATCH_ALL_TYPES(self.scalar_type(), "trace", [&] {
+    scalar_t sum = 0;
+    auto t_data = self.data_ptr<scalar_t>();
```
Comment (zou3519) [inline aten/src/ATen/native/TensorShape.cpp]: nit: Use "const auto*" to indicate it's a pointer and that we're not modifying it

File: aten/src/ATen/native/TensorShape.cpp [hunk 3]
```diff
@@ -1465,4 +1465,30 @@ Tensor& diag_out(Tensor &result, const Tensor& self, int64_t dimension) {
...
+    t_stride_0 = self.stride(0);
+    t_stride_1 = self.stride(1);
+
+    t_diag_size = std::min(self.size(0), self.size(1));
+    while(i < t_diag_size) {
+      sum += t_data[i * (t_stride_0 + t_stride_1)];
+      i++;
+    }
+
+    result.fill_(sum);
```
Comment (zou3519) [inline aten/src/ATen/native/TensorShape.cpp]: I'm wondering if `*result.data_ptr<scalar_t>() = sum` is faster

Comment (shihongzhi) [reply aten/src/ATen/native/TensorShape.cpp]: yes, you're right. I change this line

File: aten/src/ATen/native/TensorShape.cpp (hunk 1 above)
Comment (shihongzhi) [reply aten/src/ATen/native/TensorShape.cpp]: done

File: aten/src/ATen/native/TensorShape.cpp (hunk 2 above)
Comment (shihongzhi) [reply aten/src/ATen/native/TensorShape.cpp]: done

File: aten/src/ATen/native/TensorShape.cpp [hunk 4]
```diff
@@ -1469,12 +1469,12 @@ Tensor trace(const Tensor& self) {
...
   AT_DISPATCH_ALL_TYPES(self.scalar_type(), "trace", [&] {
     scalar_t sum = 0;
-    auto t_data = self.data_ptr<scalar_t>();
+    const auto* t_data = self.data_ptr<scalar_t>();
 
     int64_t i = 0;
     int64_t t_stride_0, t_stride_1, t_diag_size;
 
-    TORCH_CHECK(self.dim() == 2, "expected a matrix");
+    TORCH_CHECK(self.dim() == 2, "trace: expected a matrix, but got tensor with dim %d", self.dim());
```
Comment (zou3519) [inline aten/src/ATen/native/TensorShape.cpp]: The %d doesn't actually do anything -- TORCH_CHECK concatenates everything found after the first arg. The right way to invoke it is:
```
TORCH_CHECK(self.dim() == 2, "trace: expected a matrix, but got tensor with dim ", self.dim())
```

**Comment by shihongzhi** [comment]
> The code looks good. Could you provide a before/after performance comparison of trace on tensors with the following sizes:
> 
> * 1x1 (small)
> * 1024x1024
>   (and any others you might want)

performance benchmarks:

    import timeit
    import torch
    import itertools
    import statistics

    def test_perf(sizes, device, repeat, times):
        def _tensor(name, sizes, device):
            return '''{0} = torch.rand({1}, device="{2}");'''.format(name, sizes, device)

        setup_code = 'import torch;' + _tensor('x', sizes, device)
        test_code = '''torch.trace(x);'''
        result = timeit.repeat(setup = setup_code,stmt = test_code,repeat = repeat,number = times)
        mean = statistics.mean(result)
        std = statistics.stdev(result)
        print('''sizes = {0} std = {1} mean = {2}'''.format(sizes, std, mean))

    def test_perf_for_device(device, small, mid, large):
        print(device)
        for s in itertools.product((small, mid, large), (small, mid, large)):
            test_perf(str(s), device, 3, 300)

    test_perf_for_device("cpu", 1, 100, 1024)

pytorch:master

    cpu
    sizes = (1, 1) std = 0.000188977932464925 mean = 0.007469333000093077
    sizes = (1, 100) std = 0.00045832903142235294 mean = 0.007204115667263977
    sizes = (1, 1024) std = 0.0005223229811213363 mean = 0.00737451899961646
    sizes = (100, 1) std = 0.000229372483376989 mean = 0.007206766999540075
    sizes = (100, 100) std = 0.00019936871530362298 mean = 0.007542932001039541
    sizes = (100, 1024) std = 0.0002938848867088897 mean = 0.007327319333853666
    sizes = (1024, 1) std = 0.0001949218778043529 mean = 0.0073083453316939995
    sizes = (1024, 100) std = 7.889883887779446e-05 mean = 0.007222993997856975
    sizes = (1024, 1024) std = 0.0003551341414585057 mean = 0.00967970333294943

shihongzhi:feature/port_trace

    cpu
    sizes = (1, 1) std = 6.521108314926506e-05 mean = 0.0045181599974360625
    sizes = (1, 100) std = 6.0531971323261434e-05 mean = 0.004485020001690525
    sizes = (1, 1024) std = 0.00012657404089304485 mean = 0.004582160331968528
    sizes = (100, 1) std = 6.3884699992757e-05 mean = 0.004489179669083872
    sizes = (100, 100) std = 8.268442230894316e-05 mean = 0.004749082664299446
    sizes = (100, 1024) std = 5.5246005587137225e-05 mean = 0.004755277334576628
    sizes = (1024, 1) std = 7.378639705209506e-05 mean = 0.004510825999507991
    sizes = (1024, 100) std = 8.610108222894735e-05 mean = 0.004826352332505242
    sizes = (1024, 1024) std = 0.00021040617063372132 mean = 0.006569871999090537

File: aten/src/ATen/native/TensorShape.cpp (hunk 4 above)
Comment (shihongzhi) [reply aten/src/ATen/native/TensorShape.cpp]: sorry, it's my fault

**Comment by shihongzhi** [comment]
@zou3519 please help review the code, thanks.

File: aten/src/ATen/native/native_functions.yaml [hunk 5]
```diff
@@ -4441,7 +4441,7 @@
   use_c10_dispatcher: full
   variants: method, function
   dispatch:
-    CPU: legacy::cpu::_th_trace
+    CPU: trace
```
Comment (zou3519) [inline aten/src/ATen/native/native_functions.yaml]: Let's call this "trace_cpu" to avoid confusion

File: aten/src/ATen/native/TensorShape.cpp [hunk 6]
```diff
@@ -1465,4 +1465,30 @@ Tensor& diag_out(Tensor &result, const Tensor& self, int64_t dimension) {
   return result;
 }
 
+Tensor trace(const Tensor& self) {
+  Tensor result = at::empty({}, self.options());
+  AT_DISPATCH_ALL_TYPES(self.scalar_type(), "trace", [&] {
+    scalar_t sum = 0;
```
Comment (zou3519) [inline aten/src/ATen/native/TensorShape.cpp]: We should be performing the sum in an accumulated type, not the scalar type of `self`, to avoid precision issues. Inside the deleted TH code, we can see that it is accumulating into 'accreal'.

So this code should be:
```
// Pseudocode, possibly syntax issues
using accscalar_t = acctype<scalar_t, /*is_cuda*/True>;
```

Here's where acc_type is defined: https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/AccumulateType.h

File: aten/src/ATen/native/TensorShape.cpp [hunk 7]
```diff
@@ -1465,4 +1465,30 @@ Tensor& diag_out(Tensor &result, const Tensor& self, int64_t dimension) {
...
+    int64_t i = 0;
+    int64_t t_stride_0, t_stride_1, t_diag_size;
+
+    TORCH_CHECK(self.dim() == 2, "trace: expected a matrix, but got tensor with dim ", self.dim());
+
+    t_stride_0 = self.stride(0);
+    t_stride_1 = self.stride(1);
+
+    t_diag_size = std::min(self.size(0), self.size(1));
+    while(i < t_diag_size) {
```
Comment (zou3519) [inline aten/src/ATen/native/TensorShape.cpp]: nit: for loop looks nicer here, we save 2 lines of code

```
for (int64_t i = 0; i < t_diag_size, i++) {
...
}
```

**Review by zou3519 — CHANGES_REQUESTED** [review]
I have two requests, after that the code should be good to go

File: aten/src/ATen/native/native_functions.yaml (hunk 5 above)
Comment (shihongzhi) [reply aten/src/ATen/native/native_functions.yaml]: done

File: aten/src/ATen/native/TensorShape.cpp (hunk 6 above)
Comment (shihongzhi) [reply aten/src/ATen/native/TensorShape.cpp]: done

File: aten/src/ATen/native/TensorShape.cpp (hunk 7 above)
Comment (shihongzhi) [reply aten/src/ATen/native/TensorShape.cpp]: done

**Review by zou3519 — APPROVED** [review]
Thanks!

**Comment by anjali411** [comment]
ping @zou3519 was this ever merged?

**Comment by zou3519** [comment]
@shihongzhi Sorry, merging this fell off my radar. Could you rebase this PR? (Alternatively I can rebase this PR and fix the merge conflicts -- I don't think there will be many, but I just wanted to check in first)

**Comment by zou3519** [comment]
I shipped this as a part of https://github.com/pytorch/pytorch/pull/47126, thank you @shihongzhi for the contribution!
